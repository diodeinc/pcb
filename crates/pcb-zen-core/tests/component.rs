#[macro_use]
mod common;

snapshot_eval!(component_properties, {
    "C146731.kicad_sym" => include_str!("resources/C146731.kicad_sym"),
    "test_props.zen" => r#"
        # Import component factory from current directory.
        load(".", MyComponent = "C146731")

        # Instantiate with pin connections and a custom property.
        MyComponent(
            name = "U1",
            pins = {
                "ICLK": Net("ICLK"),
                "Q1": Net("Q1"),
                "Q2": Net("Q2"),
                "Q3": Net("Q3"),
                "Q4": Net("Q4"),
                "GND": Net("GND"),
                "VDD": Net("VDD"),
                "OE": Net("OE"),
            },
            footprint = "SMD:0805",
            properties = {"CustomProp": "Value123"},
        )
    "#
});

snapshot_eval!(interface_net_incompatible, {
    "Module.zen" => r#"
        SingleNet = interface(signal = Net)
        signal_if = SingleNet(name="sig")

        Component(
            name = "test_comp",
            footprint = "test_footprint",
            pin_defs = {"in": "1", "out": "2"},
            pins = {
                "in": signal_if,  # This should fail - interfaces not accepted for pins
                "out": Net()
            }
        )
    "#
});
