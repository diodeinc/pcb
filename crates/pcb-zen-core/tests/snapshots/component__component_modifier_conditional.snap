---
source: crates/pcb-zen-core/tests/component.rs
expression: output
---
R1 (10k) - MPN: 10K_PART Manufacturer: Vendor_10K
R2 (100k) - MPN: 100K_PART Manufacturer: Vendor_100K
R3 (1k) - MPN: None Manufacturer: None
{
    <root>: Module {
        path: <root>,
        source: "test.zen",
        children: [
            FrozenValue(
                Component {
                    name: "R1",
                    mpn: "10K_PART",
                    manufacturer: "Vendor_10K",
                    footprint: "0603",
                    prefix: "U",
                    connections: {
                        "1": FrozenValue(
                            Net {
                                name: "A",
                                id: "<ID>",
                            },
                        ),
                        "2": FrozenValue(
                            Net {
                                name: "B",
                                id: "<ID>",
                            },
                        ),
                    },
                    properties: {
                        "__resistance__": FrozenValue(
                            PhysicalValue {
                                value: 10000,
                                tolerance: 0,
                                unit: PhysicalUnitDims {
                                    current: -1,
                                    time: 0,
                                    voltage: 1,
                                    temp: 0,
                                },
                            },
                        ),
                        "resistance": FrozenValue(
                            "10k",
                        ),
                    },
                    symbol: FrozenValue(
                        Symbol {
                            name: None,
                            pins: {
                                "1": "1",
                                "2": "2",
                            },
                        },
                    ),
                },
            ),
            FrozenValue(
                Component {
                    name: "R2",
                    mpn: "100K_PART",
                    manufacturer: "Vendor_100K",
                    footprint: "0603",
                    prefix: "U",
                    connections: {
                        "1": FrozenValue(
                            Net {
                                name: "C",
                                id: "<ID>",
                            },
                        ),
                        "2": FrozenValue(
                            Net {
                                name: "D",
                                id: "<ID>",
                            },
                        ),
                    },
                    properties: {
                        "__resistance__": FrozenValue(
                            PhysicalValue {
                                value: 100000,
                                tolerance: 0,
                                unit: PhysicalUnitDims {
                                    current: -1,
                                    time: 0,
                                    voltage: 1,
                                    temp: 0,
                                },
                            },
                        ),
                        "resistance": FrozenValue(
                            "100k",
                        ),
                    },
                    symbol: FrozenValue(
                        Symbol {
                            name: None,
                            pins: {
                                "1": "1",
                                "2": "2",
                            },
                        },
                    ),
                },
            ),
            FrozenValue(
                Component {
                    name: "R3",
                    footprint: "0603",
                    prefix: "U",
                    connections: {
                        "1": FrozenValue(
                            Net {
                                name: "E",
                                id: "<ID>",
                            },
                        ),
                        "2": FrozenValue(
                            Net {
                                name: "F",
                                id: "<ID>",
                            },
                        ),
                    },
                    properties: {
                        "__resistance__": FrozenValue(
                            PhysicalValue {
                                value: 1000,
                                tolerance: 0,
                                unit: PhysicalUnitDims {
                                    current: -1,
                                    time: 0,
                                    voltage: 1,
                                    temp: 0,
                                },
                            },
                        ),
                        "resistance": FrozenValue(
                            "1k",
                        ),
                    },
                    symbol: FrozenValue(
                        Symbol {
                            name: None,
                            pins: {
                                "1": "1",
                                "2": "2",
                            },
                        },
                    ),
                },
            ),
        ],
    },
}
[]
